File: src/main/java/ru/darkkeks/vkmirror/VkMirror.java
package ru.darkkeks.vkmirror;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ru.darkkeks.vkmirror.autoreg.BotDataManager;
import ru.darkkeks.vkmirror.autoreg.VkMirrorBot;
import ru.darkkeks.vkmirror.tdlib.TdApi;
import ru.darkkeks.vkmirror.tdlib.TelegramClient;
import ru.darkkeks.vkmirror.tdlib.TelegramCredentials;
import ru.darkkeks.vkmirror.vk.ChatType;
import ru.darkkeks.vkmirror.vk.UserLongPoll;
import ru.darkkeks.vkmirror.vk.VkController;
import ru.darkkeks.vkmirror.vk.object.Message;

import javax.inject.Inject;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ScheduledExecutorService;

// TODO Think about possibility of someone else using this
//  Having a lot of accounts can be not manageable for a lot of people
//  Using one bot could be good use case if you dont mind bad look of chats
//  So a good idea would be to support both modes, but I don't feel like doing it right now, so I will just leave this
//  wall of text here, and maybe some day later I will look at it, return to this commit (#76cef1dc) and reuse a lot
//  of message sync logic.


// TODO Здесь я собираюсь попытаться описать алгоритм более абстрактно чтобы изменения старого кода не вызывали
//  |                         столько вопросов ://
//  |
//  |Два случая:
//  |  Сообщение пришло из вк
//  |    Получим объект соответствующий этому чатику из базы -- getChat()
//  |    Получим нужный клиент -- бота либо самого себя
//  |    Если это приватный чат, то:
//  |       Надо получить нужный id чата, учитывая что чат уже создан, мы уже знаем id -- это id получателя.
//  |       Отправляем сообщение
//  |    Если это групповой чат, то:
//  |       Если бота нету в чатике, то надо его добавить
//  |       Отправляем сообщение
//  |
//  |  Сообщение пришло из телеграма
//  |    ?
//  |
//  |getChat(vkPeerId):
//  |  Если объект с таким vkPeerId уже есть, достанем его из базы
//  |  Иначе
//  |    Если это приватный чат (vkPeerId -- id человека в вк):
//  |      Сделать бота -- getBotForVkUser(vkPeerId)
//  |      Поищем от моего имени бота и отправим /start, удалив сразу после этого (Боту надо либо не триггерится на
//  |             /start, либо как то помечать что бот еще не привязан и игнорить все сообщения)
//  |      Вернем объект, сохранив его в базу
//  |    Если это групповой чат (vkPeerId -- id чата привязанный ко мне):
//  |      Надо создать чатик -- createTelegramChat(vkPeerId, title) (от моего имени)
//  |      Если сообщение отправил не я, то надо добавить соответствующего бота в чатик:
//  |        Получим бота -- getBotForVkUser(message.from)
//  |      Вернем объект, сохранив его в базу
//  |
//  |getBotForVkUser()
//  |  Он будет либо уже привязан
//  |  Либо надо будет привязать нового бота к этому человеку
public class VkMirror {

    private static final Logger logger = LoggerFactory.getLogger(VkMirror.class);

    private BotDataManager botDataManager;

    private VkController vkController;
    private TelegramClient telegram;
    private VkMirrorDao vkMirrorDao;
    private ScheduledExecutorService executor;

    private Map<Integer, TelegramClient> clientsByVkId;

    private Map<Integer, Chat> chatsByPeerId;
    private Map<Integer, Chat> chatsByTelegramGroup;

    @Inject
    public VkMirror(BotDataManager botDataManager, VkController vkController, TelegramClient telegram, VkMirrorDao vkMirrorDao,
                    ScheduledExecutorService executor) {
        this.botDataManager = botDataManager;
        this.vkController = vkController;
        this.telegram = telegram;
        this.vkMirrorDao = vkMirrorDao;
        this.executor = executor;

        clientsByVkId = new HashMap<>();
        chatsByPeerId = new HashMap<>();
        chatsByTelegramGroup = new HashMap<>();
    }

    public void start() {
        telegram.onMessage(msg -> {
            if (msg.message.senderUserId == telegram.getMyId()) {
                this.executor.execute(() -> {
                    synchronized (this) {
                        onTelegramMessage(msg.message);
                    }
                });
            }
        });

        telegram.start();

        UserLongPoll longPoll = new UserLongPoll(vkController.getClient(), vkController.getActor(), this::onVkMessage);
        executor.submit(() -> vkController.run(longPoll));
    }

    /**
     * New Telegram message -> VK
     */
    private void onTelegramMessage(TdApi.Message message) {
        long chatId = message.chatId;

        if (message.senderUserId != telegram.getMyId()) return;

        TdApi.Chat chat = telegram.getChat(chatId);
        if (chat != null && chat.type instanceof TdApi.ChatTypeSupergroup) {
            int supergroupId = ((TdApi.ChatTypeSupergroup) chat.type).supergroupId;

            Chat mirrorChat = chatsByTelegramGroup.computeIfAbsent(supergroupId,
                    vkMirrorDao::getChatByTelegramGroup);

            if (mirrorChat != null) {
                logger.info("Received message from myself id={} \"{}\"", message.id, message.content);
                if (!vkMirrorDao.isSyncedTelegram(mirrorChat, message.id)) {
                    logger.info("Message {} is not synced", message.id);
                    List<Integer> ids = vkController.sendMessage(mirrorChat.getVkPeerId(), message);
                    logger.info("Produced vk messages: {}", ids.toString());
                    ids.forEach(id -> {
                        vkMirrorDao.saveMessage(mirrorChat, id, message.id);
                    });
                }
            }
        }
    }

    /**
     * VK message -> Telegram (both from me and other chat participants)
     */
    private void onVkMessage(Message message) {
        logger.info("New message from {} -> id={} \"{}\"", message.getFrom(), message.getMessageId(), message.getText());

        getTelegramChat(message.getPeerId()).thenAccept((Chat chat) -> {
            if (chat == null) {
                logger.warn("Was unable to mirror message from vk to telegram chat, getTelegramChat returned null: {}",
                        message);
                return;
            }

            // TODO Chat direction;
            if (!vkMirrorDao.isSyncedVk(chat, message.getMessageId())) {
                CompletableFuture<TelegramClient> clientFuture;

                // Message from myself
                if ((message.getFlags() & Message.Flags.OUTBOX) > 0) {
                    clientFuture = CompletableFuture.completedFuture(telegram);
                } else {
                    // Message is from one of the chat participants
                    clientFuture = getBotForVkUser(message.getFrom())
                            .thenCompose(this::getBotClient);
                }

                clientFuture.thenAccept(client -> {
                    if (chat.getType() == ChatType.PRIVATE) {
                        synchronized (this) {
                            int recipientId = client == telegram ? chat.getTelegramId() : telegram.getMyId();
                            logger.info("Sending message {}", message.getText());
                            TdApi.Message tgMessage = client.sendMessage(recipientId, message.getText()).join();
                            logger.info("Message sent {}", tgMessage);
                            vkMirrorDao.saveMessage(chat, message.getMessageId(), tgMessage.id);
                            logger.info("Synced message {}!", tgMessage);
                        }
                    } else if (chat.getType() == ChatType.GROUP) {
                        client.groupById(chat.getTelegramId()).thenAccept(group -> {
                            logger.info("Sending to chat {}", group.id);
                            executor.submit(() -> {
                                synchronized (this) {
                                    TdApi.Message tgMessage = client.sendMessage(group.id, message.getText()).join();
                                    logger.info("Message sent {}", tgMessage);
                                    vkMirrorDao.saveMessage(chat, message.getMessageId(), tgMessage.id);
                                    logger.info("Synced message {}!", tgMessage);
                                }
                            });
                        });
                    }
                });
            }
        });
    }

    /**
     * Chat has to be from my point of view, otherwise every private chat would have the same id
     * But to send message from bot we need chat from bots' perspective :(
     * <p>
     * Lets get all chats from my point of view and apply some if-magic to it afterwards ?
     */
    private CompletableFuture<Chat> getTelegramChat(int vkPeerId) {
        Chat chat = chatsByPeerId.computeIfAbsent(vkPeerId, vkMirrorDao::getChatByVkPeer);

        if (chat != null) {
            return CompletableFuture.completedFuture(chat);

            // TODO Fix this check for groups and private chats
//            return telegram.groupById(chat.getTelegramId()).thenApply(group -> {
//                if (group == null) {
//                    logger.error("Saved group does not exist (Probably user left)");
//                }
//
//                return chat;
//            });
        } else if (vkController.isMultichat(vkPeerId)) {
            return CompletableFuture.supplyAsync(() -> {
                TdApi.Chat tgChat = createTelegramGroup(vkPeerId, vkController.getChannelTitle(vkPeerId)).join();
                TdApi.ChatTypeSupergroup supergroup = (TdApi.ChatTypeSupergroup) tgChat.type;

                Chat mirrorChat = Chat.groupChat(vkPeerId, supergroup.supergroupId);

                chatsByPeerId.put(vkPeerId, mirrorChat);
                chatsByTelegramGroup.put(supergroup.supergroupId, mirrorChat);

                vkMirrorDao.save(mirrorChat);
                return mirrorChat;
            });
        } else if (vkController.isPrivateChat(vkPeerId)) {
            return createPrivateChat(vkPeerId).thenApply(botClient -> {
                if (botClient == null) {
                    return null;
                }

                Chat mirrorChat = Chat.privateChat(vkPeerId, botClient.getMyId());

                chatsByPeerId.put(vkPeerId, mirrorChat);

                vkMirrorDao.save(mirrorChat);
                return mirrorChat;
            });
        }

        throw new UnsupportedOperationException("Group chats are not supported yet");
    }

    private CompletableFuture<TdApi.Chat> createTelegramGroup(int vkPeerId, String title) {
        String chatUrl = vkController.getChatUrl(vkPeerId);
        return telegram.createChannel(title, chatUrl).thenApply(chat -> {
            TdApi.ChatType type = chat.type;
            if (type instanceof TdApi.ChatTypeSupergroup) {
                // TODO Should join every person that is already bound to a bot;
                //  Other people should be added only on activity in chat
//                telegram.chatAddUser(chat.id, bot.getMyId()).join();

                Path photo = vkController.downloadConversationImage(vkPeerId);
                if (photo != null) {
                    telegram.setGroupPhoto(chat.id, photo).join();
                }

            } else {
                throw new IllegalStateException("New chat is not channel");
            }
            return chat;
        });
    }

    private CompletableFuture<TelegramClient> createPrivateChat(int vkPeerId) {
        return getBotForVkUser(vkPeerId).thenCompose(bot -> {
            if (bot == null) {
                return CompletableFuture.completedFuture(null);
            }

            return telegram.searchPublicUsername(bot.getUsername()).thenCompose(chat -> {
                // We assume bot will delete this message for both of us
                telegram.sendMessage(chat.id, "/start");
                return getBotClient(bot);
            });
        });
    }

    private CompletableFuture<VkMirrorBot> getBotForVkUser(int vkId) {
        return botDataManager.getBotForVk(vkId);
    }

    private CompletableFuture<TelegramClient> getBotClient(VkMirrorBot bot) {
        if (clientsByVkId.containsKey(bot.getVkId())) {
            return CompletableFuture.completedFuture(clientsByVkId.get(bot.getVkId()));
        } else {
            return CompletableFuture.supplyAsync(() -> {
                TelegramClient client = new TelegramClient(TelegramCredentials.bot(
                        Config.API_ID, Config.API_HASH, bot.getToken()));

                // TODO Test if bot can actually revoke this for me
                client.onMessage(newMessage -> {
                    TdApi.Message message = newMessage.message;
                    if (message.content instanceof TdApi.MessageText) {
                        TdApi.FormattedText text = ((TdApi.MessageText) message.content).text;
                        if (text.text.equals("/start")) {
                            client.deleteMessage(message.chatId, true, message.id).exceptionally(e -> {
                                logger.error("Failed to delete /start message", e);
                                return null;
                            });
                        }
                    }
                });

                client.start();

                clientsByVkId.put(bot.getVkId(), client);
                return client;
            });
        }
    }
}

File: src/main/java/ru/darkkeks/vkmirror/VkMirror.kt
package ru.darkkeks.vkmirror

import kotlinx.coroutines.*
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.kodein.di.Kodein
import org.kodein.di.generic.instance
import org.litote.kmongo.Id
import org.litote.kmongo.newId
import ru.darkkeks.vkmirror.bot.MirrorBot
import ru.darkkeks.vkmirror.bot.START
import ru.darkkeks.vkmirror.tdlib.TelegramClient
import ru.darkkeks.vkmirror.tdlib.botTelegramCredentials
import ru.darkkeks.vkmirror.tdlib.internal.TdApi
import ru.darkkeks.vkmirror.util.createLogger
import ru.darkkeks.vkmirror.vk.*
import ru.darkkeks.vkmirror.vk.`object`.Message
import java.util.concurrent.ConcurrentHashMap

class VkMirror(kodein: Kodein) {

    private val vk: VkController by kodein.instance()
    private val telegram: TelegramClient by kodein.instance()

    private val dao = VkMirrorDao(kodein)

    private val clients = ConcurrentHashMap<MirrorBot, TelegramClient>()

    private val chatLocks = ConcurrentHashMap<Id<Chat>, Mutex>()

    private val scope = CoroutineScope(SupervisorJob())

    suspend fun start() {
        telegram.subscribe<TdApi.UpdateNewMessage> {
            handleTelegramMessage(it.message)
        }

        telegram.subscribe<TdApi.UpdateMessageSendSucceeded> {
            handleTelegramMessage(it.message)
        }

        telegram.subscribe<TdApi.UpdateChatReadInbox> {
            logger.info("Received UpdateChatReadInbox in chat ${it.chatId} with last message ${it.lastReadInboxMessageId}")
            scope.launch {
                val chat = telegramChatIdToMirrorChat(it.chatId) ?: return@launch
                val message = telegram.getMessage(it.chatId, it.lastReadInboxMessageId)
                val messageIds = dao.getMessageLinks(chat._id, telegram.myId, it.lastReadInboxMessageId)
                val idFromBotPerspective = messageIds[message.senderUserId] ?: return@launch
                val vkMessageId = dao.getSyncedMessageByTelegramId(chat._id, idFromBotPerspective)?.vkId
                        ?: return@launch
                logger.info("Marking vk messages up to $vkMessageId in peer ${chat.peerId} as read")
                vk.markAsRead(chat.peerId, vkMessageId)
            }
        }

        telegram.start()
        vk.runLongPoll(VkListener())
    }

    private fun handleTelegramMessage(message: TdApi.Message) {
        if (message.sendingState != null) return

        logger.info("Incoming message from chat ${message.chatId} with id ${message.id}")
        scope.launch {
            val mirrorChat = telegramChatIdToMirrorChat(message.chatId) ?: return@launch

            launch {
                addMessageIdLink(mirrorChat, telegram.myId, message)
            }

            if (message.senderUserId == telegram.myId) {
                logger.info("Received message from myself id=${message.id} \"${message.content}\"")
                if (!dao.isSyncedTelegram(mirrorChat, message.id)) {
                    val vkMessages = vk.adaptMessage(mirrorChat.peerId, message)
                    getChatLock(mirrorChat).withLock {
                        if (!dao.isSyncedTelegram(mirrorChat, message.id)) {
                            logger.info("Message {} is not synced", message.id)

                            val ids = vk.sendMessages(vkMessages)
                            logger.info("Produced vk messages: {}", ids.toString())

                            dao.saveVkMessages(mirrorChat, ids, message.id)
                        }
                    }
                }
            }
        }
    }

    private suspend fun addMessageIdLink(chat: Chat, userId: Int, message: TdApi.Message) {
        logger.info("Adding message-id link for chat ${chat.peerId} for user $userId, messageId ${message.id}")
        val link = MessageIdLink(newId(), chat._id, message.content.toString(), message.date.toLong(), userId, message.id)
        dao.saveMessageIdLink(link)
    }

    private suspend fun telegramChatIdToMirrorChat(chatId: Long, client: TelegramClient = telegram): Chat? {
        val chat = telegram.getChat(chatId) ?: return null

        val id = when (val type = chat.type) {
            is TdApi.ChatTypePrivate -> {
                if (client.myId != telegram.myId) {
                    client.myId
                } else {
                    type.userId
                }
            }
            is TdApi.ChatTypeSupergroup -> type.supergroupId
            else -> return null
        }

        return dao.getChatByTelegramId(id)
    }

    private suspend fun mirrorChatForPeerId(peerId: Int): Chat? {
        return dao.getChatByVkId(peerId) ?: when {
            isPrivateChat(peerId) -> {
                // TODO Add logging in places like this ?
                val bot = dao.getBotByVkId(peerId) ?: return null

                val chat = telegram.searchPublicUsername(bot.username)
                telegram.sendMessage(chat.id, START)

                // Just so we delete /start message
                val client = createBotClient(bot)

                Chat.privateChat(peerId, client.myId)
            }
            isMultichat(peerId) -> {
                val tgChat = createTelegramGroup(peerId)
                val supergroup = tgChat.type as TdApi.ChatTypeSupergroup

                Chat.groupChat(peerId, supergroup.supergroupId)
            }
            else -> throw UnsupportedOperationException("Group chats are not supported yet")
        }.also {
            dao.saveMirrorChat(it)
        }
    }

    private suspend fun createTelegramGroup(peerId: Int): TdApi.Chat {
        val chat = telegram.createSupergroup(vk.getChatTitle(peerId), getChatUrl(peerId))


        coroutineScope {
            launch {
                val photo = vk.downloadConversationImage(peerId)
                if (photo != null) {
                    telegram.setGroupPhoto(chat.id, photo)
                }
            }

            vk.getMultichatMembers(peerId).mapNotNull {
                dao.getBotByVkId(it)
            }.forEach {
                launch {
                    telegram.chatAddUser(chat.id, createBotClient(it).myId)
                }
            }
        }

        return chat
    }

    private suspend fun createBotClient(bot: MirrorBot): TelegramClient {
        return clients.computeIfAbsent(bot) {
            TelegramClient(botTelegramCredentials(API_ID, API_HASH, bot.token)).also { client ->
                client.subscribe<TdApi.UpdateNewMessage> { update ->
                    val message = update.message
                    logger.info("Bot incoming message from chat ${message.chatId} with id ${message.id}")

                    scope.launch {
                        val chat = telegramChatIdToMirrorChat(message.chatId, client = client) ?: return@launch
                        addMessageIdLink(chat, client.myId, message)
                    }

                    when (val content = message.content) {
                        is TdApi.MessageText -> {
                            if (content.text.text == "/start") {
                                scope.launch {
                                    // TODO Log errors ?
                                    client.deleteMessage(message.chatId, true, message.id)
                                }
                            }
                        }
                    }
                }
            }
        }.apply {
            start()
        }
    }

    private fun getChatLock(chat: Chat): Mutex {
        return chatLocks.computeIfAbsent(chat._id) { Mutex() }
    }

    private suspend fun sendTypingAction(vkUserId: Int, chatId: suspend (TelegramClient) -> Long) {
        val bot = dao.getBotByVkId(vkUserId)
        if (bot == null) {
            logger.info("Received typing status for user without a bot")
        } else {
            val client = createBotClient(bot)
            client.sendChatAction(chatId(client), TdApi.ChatActionTyping())
        }
    }

    inner class VkListener : UserLongPollListener {

        private suspend fun sendVkMessage(client: TelegramClient, chat: Chat, chatId: Long, message: Message) {
            logger.info("Sending to chat $chatId")

            val messages = vk.adaptMessage(message)
            messages.forEach { inputContent ->
                val tgMessage = client.sendMessage(chatId, inputContent)
                logger.info("Message sent $tgMessage")
                addMessageIdLink(chat, client.myId, tgMessage)
                dao.saveTelegramMessages(chat, message.messageId, listOf(tgMessage.id), client.myId)
            }
        }

        override fun newMessage(message: Message) {
            scope.launch {
                logger.info("New message from ${message.from} -> id=${message.messageId} \"${message.text}\"")

                val chat = mirrorChatForPeerId(message.peerId) ?: return@launch

                getChatLock(chat).withLock {
                    if (dao.isSyncedVk(chat, message.messageId)) return@launch

                    val client = if (message.flags and Message.Flags.OUTBOX > 0) {
                        telegram
                    } else {
                        val bot = dao.getBotByVkId(message.from) ?: return@launch
                        createBotClient(bot)
                    }

                    when (chat.type) {
                        ChatType.PRIVATE -> {
                            val recipient = if (client == telegram) chat.telegramId else telegram.myId
                            sendVkMessage(client, chat, recipient.toLong(), message)
                        }
                        ChatType.GROUP -> {
                            val groupId = chat.telegramId
                            val group = client.groupById(groupId)

                            if (group == null) {
                                val newGroup = telegram.groupById(groupId)
                                        ?: throw IllegalStateException("User left group") // TODO: Handle leaving group
                                telegram.chatAddUser(newGroup.id, client.myId)

                                val joinedGroup = client.groupById(groupId)
                                        ?: throw IllegalStateException("Failed to add bot to group") // TODO: Handle add failure

                                sendVkMessage(client, chat, joinedGroup.id, message)
                            } else {
                                sendVkMessage(client, chat, group.id, message)
                            }
                        }
                    }
                }
            }
        }

        override fun editMessage(editedMessage: Message) {

        }

        override fun readUpToInbound(event: MessageReadUpTo) {
            scope.launch {
                val chat = mirrorChatForPeerId(event.peerId) ?: return@launch
                getChatLock(chat).withLock {
                    val message = dao.getSyncedMessageByVkId(chat._id, event.messageId) ?: return@launch
                    val messageIds = dao.getMessageLinks(chat._id, message.sender, message.telegramId)
                    val messageIdFromByPerspective = messageIds[telegram.myId] ?: return@launch
                    val telegramChatId = when (chat.type) {
                        ChatType.PRIVATE -> chat.telegramId.toLong()
                        ChatType.GROUP -> {
                            val supergroup = telegram.groupById(chat.telegramId) ?: return@launch
                            supergroup.id
                        }
                    }
                    telegram.readMessages(telegramChatId, messageIdFromByPerspective, forceRead = true)
                }
            }
        }

        override fun readUpToOutbound(event: MessageReadUpTo) {
            // Управлять прочтением собщений ботом невозможно
        }

        override fun userIsTyping(userId: Int) {
            scope.launch {
                sendTypingAction(userId) {
                    telegram.myId.toLong()
                }
            }
        }

        override fun userIsTyping(event: UserIsTyping) {
            scope.launch {
                val peerId = if (isGroup(event.chatId)) event.chatId else event.chatId + MULTICHAT_BASE
                val chat = mirrorChatForPeerId(peerId) ?: return@launch

                sendTypingAction(event.userId) { client ->
                    client.groupById(chat.telegramId)?.id ?: throw IllegalStateException("Bot is not in group")
                }
            }
        }

        override fun usersAreTyping(event: UsersAreTyping) {
            scope.launch {
                val peerId = event.peerId
                val chat = mirrorChatForPeerId(peerId) ?: return@launch

                event.userIds.forEach { userId ->
                    launch {
                        sendTypingAction(userId) { client ->
                            client.groupById(chat.telegramId)?.id ?: throw IllegalStateException("Bot is not in group")
                        }
                    }
                }
            }
        }
    }

    companion object {
        val logger = createLogger<VkMirror>()
    }
}